tools/clang/lib/SPIRV/EmitVisitor.cpp

```diff
@@ -808,5 +826,188 @@ bool EmitVisitor::visit(SpirvVectorShuffle *inst) {
   return true;
 }
 
+// EmitTypeHandler ------
+
+void EmitTypeHandler::initTypeInstruction(spv::Op op) {
+  curTypeInst.clear();
+  curTypeInst.push_back(static_cast<uint32_t>(op));
+}
+
+void EmitTypeHandler::finalizeTypeInstruction() {
+  curTypeInst[0] |= static_cast<uint32_t>(curTypeInst.size()) << 16;
+  typeConstantBinary->insert(typeConstantBinary->end(), curTypeInst.begin(),
+                             curTypeInst.end());
+}
+
+uint32_t EmitTypeHandler::getResultIdForType(const SpirvType *type,
+                                             SpirvLayoutRule rule,
+                                             bool *alreadyExists) {
+  // Check if the type has already been emitted.
+  auto foundType = emittedTypes.find(type);
+  if (foundType != emittedTypes.end()) {
+    auto foundLayoutRule = foundType->second.find(rule);
+    if (foundLayoutRule != foundType->second.end()) {
+      *alreadyExists = true;
+      return foundLayoutRule->second;
+    }
+  }
+
+  *alreadyExists = false;
+  const uint32_t id = module.takeNextId();
+  emittedTypes[type][rule] = id;
+  return id;
+}
+
+uint32_t EmitTypeHandler::emitType(const SpirvType *type,
+                                   SpirvLayoutRule rule) {
+  //
+  // TODO: This method is currently missing decorations for types completely.
+  //
+
+  bool alreadyExists = false;
+  const uint32_t id = getResultIdForType(type, rule, &alreadyExists);
+
+  // If the type has already been emitted, we just need to return its
+  // <result-id>.
+  if (alreadyExists)
+    return id;
+
+  if (isa<VoidType>(type)) {
+    initTypeInstruction(spv::Op::OpTypeVoid);
+    curTypeInst.push_back(id);
+    finalizeTypeInstruction();
+  }
```

Comment:
Better to do `else if` so at run time we don't need to check all of the cases for all the time.